From 87c74f69dc75e94b64a258452c7cb22a421c5f69 Mon Sep 17 00:00:00 2001
From: Alex Robinson <arobinson@3rdeyecam.com>
Date: Wed, 5 Aug 2020 14:17:50 -0500
Subject: [PATCH] showing unlock page re-orients the screen

---
 src/lockscreen.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/src/lockscreen.c b/src/lockscreen.c
index e428f6a..3a486df 100644
--- a/src/lockscreen.c
+++ b/src/lockscreen.c
@@ -20,6 +20,7 @@
 #include <glib/gi18n.h>
 #include <math.h>
 #include <time.h>
+#include <shell.h>
 
 #define HANDY_USE_UNSTABLE_API
 #include <handy.h>
@@ -402,6 +403,7 @@ carousel_position_notified_cb (PhoshLock
                                GParamSpec      *pspec,
                                HdyCarousel     *carousel)
 {
+  PhoshShell *shell = phosh_shell_get_default ();
   PhoshLockscreenPrivate *priv = phosh_lockscreen_get_instance_private (self);
   double position;
 
@@ -418,6 +420,7 @@ carousel_position_notified_cb (PhoshLock
       priv->idle_timer = g_timeout_add_seconds (LOCKSCREEN_IDLE_SECONDS,
                                                 (GSourceFunc) keypad_check_idle,
                                                 self);
+      phosh_shell_rotate_display (shell, 0);
     }
 
     return;
