From 63694fbfb1db22abcffaa0c1dbf9be416389c258 Mon Sep 17 00:00:00 2001
From: Alex Robinson <arobinson@3rdeyecam.com>
Date: Wed, 5 Aug 2020 14:31:32 -0500
Subject: [PATCH] lock rotation on lockscreen

---
 src/rotation.c | 25 +++++++++++++++++++++++++
 1 file changed, 25 insertions(+)

diff --git a/src/rotation.c b/src/rotation.c
index b3609ee..b474d51 100644
--- a/src/rotation.c
+++ b/src/rotation.c
@@ -108,6 +108,10 @@ on_has_accelerometer_changed (PhoshRotation           *self,
 {
   gboolean has_accel;
 
+  /* Don't claim if locked, no rotation needed */
+  if (phosh_lockscreen_manager_get_locked(self->lockscreen_manager))
+    return;
+
   has_accel = phosh_dbus_sensor_proxy_get_has_accelerometer (
     PHOSH_DBUS_SENSOR_PROXY (self->sensor_proxy_manager));
 
@@ -115,6 +119,22 @@ on_has_accelerometer_changed (PhoshRotation           *self,
   phosh_rotation_claim_accelerometer (self, has_accel);
 }
 
+static void
+on_lockscreen_manager_locked (PhoshRotation *self, GParamSpec *pspec,
+                              PhoshLockscreenManager *lockscreen_manager)
+{
+  PhoshShell *shell = phosh_shell_get_default ();
+  gboolean locked;
+
+  phosh_shell_rotate_display (shell, 0);
+  
+  g_return_if_fail (PHOSH_IS_ROTATION (self));
+  g_return_if_fail (PHOSH_IS_LOCKSCREEN_MANAGER (lockscreen_manager));
+
+  locked = phosh_lockscreen_manager_get_locked(self->lockscreen_manager);
+  phosh_rotation_claim_accelerometer (self, !locked);
+}
+
 static void
 on_accelerometer_orientation_changed (PhoshRotation          *self,
                                       GParamSpec              *pspec,
@@ -191,6 +211,11 @@ static void
 phosh_rotation_constructed (GObject *object)
 {
   PhoshRotation *self = PHOSH_ROTATION (object);
+  g_signal_connect_swapped (self->lockscreen_manager,
+                            "notify::locked",
+                            (GCallback) on_lockscreen_manager_locked,
+                            self);
+
   g_signal_connect_swapped (self->sensor_proxy_manager,
                             "notify::accelerometer-orientation",
                             (GCallback) on_accelerometer_orientation_changed,
