diff --git a/src/proximity.c b/src/proximity.c
index 0beca69..2d629a6 100644
--- a/src/proximity.c
+++ b/src/proximity.c
@@ -7,6 +7,7 @@
  */
 
 #define G_LOG_DOMAIN "phosh-proximity"
+#define NEAR_POWER_TIMEOUT 5 /* seconds */
 
 #include "config.h"
 #include "fader.h"
@@ -34,6 +35,8 @@ typedef struct _PhoshProximity {
   GObject parent;
 
   gboolean claimed;
+  gint near_timeoutid;
+  int ps_mode;
   PhoshSensorProxyManager *sensor_proxy_manager;
   PhoshLockscreenManager *lockscreen_manager;
   PhoshFader *fader;
@@ -141,6 +144,12 @@ on_lockscreen_manager_locked (PhoshProximity *self, GParamSpec *pspec,
   phosh_proximity_claim_proximity (self, !locked);
 }
 
+static gboolean
+on_near_timeout (PhoshProximity *self)
+{
+  ps_mode = PHOSH_MONITOR_POWER_SAVE_MODE_OFF;
+  return FALSE;
+}
 
 static void
 on_proximity_near_changed (PhoshProximity          *self,
@@ -148,17 +157,28 @@ on_proximity_near_changed (PhoshProximity          *self,
                            PhoshSensorProxyManager *sensor)
 {
   gboolean near;
-  int ps_mode;
   PhoshShell *shell = phosh_shell_get_default ();
-  PhoshMonitor *monitor = phosh_shell_get_builtin_monitor (shell);  
-  
+  PhoshMonitor *monitor = phosh_shell_get_builtin_monitor (shell);
+  PhoshWayland *wl = phosh_wayland_get_default ();
+
   near = phosh_dbus_sensor_proxy_get_proximity_near (
     PHOSH_DBUS_SENSOR_PROXY (self->sensor_proxy_manager));
 
   g_debug ("Proximity near changed: %d", near);
   if (near) {
-      ps_mode = PHOSH_MONITOR_POWER_SAVE_MODE_OFF;
+    if (!self->fader) {
+      self->fader = phosh_fader_new (phosh_wayland_get_zwlr_layer_shell_v1 (wl),
+                                     monitor->wl_output);
+      gtk_widget_show (GTK_WIDGET (self->fader));
+      near_timeoutid = NEAR_POWER_TIMEOUT;
+      if (near_timeoutid > 0) {
+        g_timeout_add_seconds (NEAR_POWER_TIMEOUT,
+                              (GSourceFunc)on_near_timeout,
+                              self);
+      }  
+    }
   } else {
+      g_clear_pointer (&self->fader, phosh_cp_widget_destroy);
       ps_mode = PHOSH_MONITOR_POWER_SAVE_MODE_ON;
   }
   phosh_monitor_set_power_save_mode (monitor, ps_mode);
